prow/git/git_command_test.go

```diff
@@ -17,99 +17,257 @@ limitations under the License.
+			simulateExecFailures: 4,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+			},
+			err: true,
 		},
 		{
-			commandRe: g + " clone /[^ ]+/org/repo.git /[^ ]+",
+			name:                 "happy path with GHE",
+			base:                 "https://corp.github.initech.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 0,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@corp.github.initech.com/org/repo", tempdir.recorded(8) + "/org/repo.git"},
+				{"", gitbin, "clone", tempdir.recorded(8) + "/org/repo.git", tempdir.recorded(9)},
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(9),
+				git:  gitbin,
+				base: urlMustParse("https://corp.github.initech.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
 		},
-	},
-		T: t,
 	}
-	execWithDir = executor.execWithDir
 	defer func() { execWithDir = defaultExecWithDir }()
-	base, _ := url.Parse("https://github.com")
-	savedBase := *base
-	client, err := NewClient(base)
-	if err != nil {
-		t.Errorf("Unexpected error %v", err)
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			simulatedFailures := 0
+			var actualCommands [][]string
+			execWithDir = func(dir, name string, args ...string) ([]byte, error) {
+				actualCommands = append(actualCommands, append([]string{dir, name}, args...))
+				if simulatedFailures < tc.simulateExecFailures {
+					simulatedFailures++
+					return nil, fmt.Errorf("Simulated failure %d", simulatedFailures)
+				}
+				return nil, nil
+			}
+			baseURL, err := url.Parse(tc.base)
+			if err != nil {
+				t.Errorf("Unexpected error parsing url %s: %v", tc.base, err)
+			}
+			client, err := NewClient(baseURL)
+			if err != nil {
+				t.Errorf("Unexpected error creating Client: %v", err)
+			}
+			client.SetCredentials(tc.user, tc.tokenGenerator)
+
+			actual, err := client.Clone(tc.repo)
+			switch {
+			case !tc.err && err != nil:
+				t.Errorf("Unexpected error result: %v", err)
+			case tc.err && err == nil:
+				t.Errorf("Expected error, got nil")
+			}
+			if tc.err && actual != nil {
+				t.Errorf("Unexpected non-nil result")
+			}
+			if !tc.err {
+				// check actual == expected
+				tc.expected.logger = client.logger
+				if !reflect.DeepEqual(&tc.expected, actual) {
+					t.Errorf("Unexpected value, diff: %s", diff.ObjectReflectDiff(actual, &tc.expected))
```

Comment:
diff.ObjectReflectDiff doesn't work well with unexported fields, it is pretty specifically designed for k8s resources, which Repo is not. (I assume this is why you're not using semantically equal too)

Reply:
oh really? ok I will use ObjectGoPrintSideBySide then. At least it is printing unexported fields in readable form.

Reply:
I could be wrong, I just had trouble with it recently